fe1-web/src/features/lao/network/LaoGreetWatcher.ts

```diff
@@ -0,0 +1,104 @@
+import { Store } from 'redux';
+
+import { getNetworkManager } from 'core/network';
+import { getMessagesState } from 'core/network/ingestion';
+import { ExtendedMessage } from 'core/network/ingestion/ExtendedMessage';
+import { ActionType, ObjectType } from 'core/network/jsonrpc/messages';
+import { PublicKey } from 'core/objects';
+import { dispatch } from 'core/redux';
+import { WitnessMessage } from 'features/witness/network/messages';
+
+import { Server } from '../objects/Server';
+import { addServer } from '../reducer/ServerReducer';
+import { GreetLao } from './messages/GreetLao';
+
+export const handleLaoGreet = (greetLaoMsg: GreetLao, publicKey: PublicKey) => {
+  dispatch(
+    addServer(
+      new Server({
+        laoId: greetLaoMsg.lao,
+        address: greetLaoMsg.address,
+        serverPublicKey: publicKey,
+        frontendPublicKey: greetLaoMsg.frontend,
+      }).toState(),
+    ),
+  );
+
+  // connect to all received peer addresses
+  // after connecting to each peer, we will send a catchup which contains the lao#create message
+  // as soon as we have parsed this message, we will connect to the LAO on the new connection as well
+  // which will trigger another lao#greet message
+  // IMPORTANT: The network manager deduplicates connections to the same address (string)
+  // and the received peer addresses are supposed to be the canonical ones.
+  // Hence we just have to make sure that the first connection is also to the canonical
+  // address, otherwise a client will connect to the same server twice (e.g. using its IP and then
+  // then using the canonical domain address)
+  const networkManager = getNetworkManager();
+  for (const peerAddress of greetLaoMsg.peers) {
+    networkManager.connect(peerAddress.address);
+  }
+};
+
+/**
+ * Watches the redux store for new message#witness messages for lao#greet messages since they only
+ * become valid after they are witnessed by the corresponding frontend
+ * @remark Implemented analogous to 'makeMessageStoreWatcher'
+ * @param store The redux store to watch
+ * @param laoGreetSignatureHandler The function to call when a signature is added to a lao#greet message
+ */
+export const makeLaoGreetStoreWatcher = (
```

Comment:
Yes that indeed sounds like much better solution, thank you 😁

Reply:
I now additionally created a reducer that just stored the ids of greeting messages. This allows the watcher to operate more efficiently; otherwise we would need to look through all messages all the time